spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnSuite.scala

```diff
@@ -124,6 +130,251 @@ trait IdentityColumnSuiteBase extends IdentityColumnTestUtils {
+        sql(s"SELECT COUNT(DISTINCT id) FROM delta.`$v5TablePath`"),
+        Row(4L)
+      )
+
+      val deltaLog = DeltaLog.forTable(spark, v5TablePath)
+      val protocolBeforeUpdate = deltaLog.snapshot.protocol
+
+      // ALTER TABLE should drop the IDENTITY columns and keeps the protocol version unchanged.
+      sql(s"ALTER TABLE delta.`$v5TablePath` ADD COLUMNS (value2 DOUBLE)")
+      deltaLog.update()
+      assert(deltaLog.snapshot.protocol == protocolBeforeUpdate)
+      assert(!ColumnWithDefaultExprUtils.hasIdentityColumn(deltaLog.snapshot.schema))
+
+      // Specifying a min writer version should not enable IDENTITY column.
+      sql(s"ALTER TABLE delta.`$v5TablePath` SET TBLPROPERTIES ('delta.minWriterVersion'='4')")
+      deltaLog.update()
+      assert(deltaLog.snapshot.protocol == Protocol(1, 4))
+      assert(!ColumnWithDefaultExprUtils.hasIdentityColumn(deltaLog.snapshot.schema))
+    }
+  }
+
+  for {
+    generatedAsIdentityType <- GeneratedAsIdentityType.values
+  } {
+    test(
+        "replace table with identity column should upgrade protocol, "
+          + s"identityType: $generatedAsIdentityType") {
+      def getProtocolVersions: (Int, Int) = {
+        sql(s"DESC DETAIL $tblName")
+          .select("minReaderVersion", "minWriterVersion")
+          .as[(Int, Int)]
+          .head()
+      }
+
+      withTable(tblName) {
+        createTable(
+          tblName,
+          Seq(
+            TestColumnSpec(colName = "id", dataType = LongType),
+            TestColumnSpec(colName = "value", dataType = IntegerType))
+        )
+        assert(getProtocolVersions == (1, 2) || getProtocolVersions == (2, 5))
+        assert(DeltaLog.forTable(spark, TableIdentifier(tblName)).snapshot.version == 0)
+
+        replaceTable(
+          tblName,
+          Seq(
+            IdentityColumnSpec(
+              generatedAsIdentityType,
+              startsWith = Some(1),
+              incrementBy = Some(1)
+            ),
+            TestColumnSpec(colName = "value", dataType = IntegerType)
+          )
+        )
+        assert(getProtocolVersions == (1, 6) || getProtocolVersions == (2, 6))
+        assert(DeltaLog.forTable(spark, TableIdentifier(tblName)).snapshot.version == 1)
+      }
+    }
+  }
+
+  test("identity value start at boundaries") {
+    val starts = Seq(Long.MinValue, Long.MaxValue)
+    val steps = Seq(1, 2, -1, -2)
+    for {
+      generatedAsIdentityType <- GeneratedAsIdentityType.values
+      start <- starts
+      step <- steps
+    } {
+      withTable(tblName) {
+        createTableWithIdColAndIntValueCol(
+          tblName, generatedAsIdentityType, Some(start), Some(step))
+        val table = DeltaLog.forTable(spark, TableIdentifier(tblName))
+        val actualSchema =
+          DeltaColumnMapping.dropColumnMappingMetadata(table.snapshot.metadata.schema)
+        assert(actualSchema === expectedSchema(generatedAsIdentityType, start, step))
+        if ((start < 0L) == (step < 0L)) {
```

Comment:
This could use a quick comment about Long underflow and overflow. It was not obvious, especially because the exception message isn't checked.